prgenv-gnu with ROCm 7

msimberg opened the pull request.

msimberg committed:
Add rccl and rccl-tests to rocm uenv

msimberg committed:
Disable duplicates strategy for now

msimberg committed:
Verify that we can't concretize with one cray-mpich

msimberg committed:
Merge remote-tracking branch 'origin/main' into prgenv-gnu-rocm-rccl-tests

msimberg committed:
Enable duplicates:strategy:full again

msimberg committed:
Add prgenv-gnu rocm 7.0.2 uenv

msimberg committed:
Update spack-packages

msimberg committed:
Temporarily disable rocprofiler-systems

msimberg commented:
ROCm is a menace: https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/551234120955960/1440398897047560/-/jobs/12201564988#L3724. hipblaslt seems to be picking up amdclang++ from the system... needs further investigation.

msimberg committed:
Merge remote-tracking branch 'origin/main' into prgenv-gnu-rocm-7

msimberg committed:
Disable hipblaslt for now

msimberg committed:
Fix linaro forge config

msimberg committed:
Disable tensile temporarily

msimberg commented:
cscs-ci run alps;system=beverin;uarch=mi200;uenv=prgenv-gnu:25.12

msimberg committed:
Apply suggestion from @msimberg

msimberg commented:
> Hi, Could you add rocprofiler_sdk/package.py to the recipe ? needed for linaro...

I'll try it out, hopefully no bigger issues (though note that I have issues with other packages before this is usable unfortunately).

iomaganaris committed:
Try ROCm 7.1.0 from spack PR

iomaganaris committed:
Trying to fix hipblaslt

afzpatel commented:
> > ROCm is a menace: https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/551234120955960/1440398897047560/-/jobs/12201564988#L3724. hipblaslt seems to be picking up amdclang++ from the system... needs further investigation.
> 
> Yes, it looks like an older version of ROCm is installed on the system and it's choosing the incorrect version of amdclang++. I'll see if I can reproduce the issue and put in a fix.

I was able to reproduce and fix the previous error with hipblaslt by setting Tensile_COMPILER in the recipe:
```
diff --git a/repos/spack_repo/builtin/packages/hipblaslt/package.py b/repos/spack_repo/builtin/packages/hipblaslt/package.py
index eba3fbc25b..e2337352ee 100644
--- a/repos/spack_repo/builtin/packages/hipblaslt/package.py
+++ b/repos/spack_repo/builtin/packages/hipblaslt/package.py
@@ -278,6 +278,11 @@ class Hipblaslt(CMakePackage):
                     "ROCROLLER_ASSEMBLER_PATH", f"{self.spec['llvm-amdgpu'].prefix}/bin/amdclang++"
                 )
             )
+            args.append(
+                self.define(
+                    "Tensile_COMPILER", f"{self.spec['llvm-amdgpu'].prefix}/bin/amdclang++"
+                )
+            )
         if self.spec.satisfies("@7.1:"):
             args.append(self.define("HIPBLASLT_ENABLE_CLIENT", self.run_tests))
         else:
```

afzpatel commented:
> > > > ROCm is a menace: https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/551234120955960/1440398897047560/-/jobs/12201564988#L3724. hipblaslt seems to be picking up amdclang++ from the system... needs further investigation.
> > > 
> > > 
> > > Yes, it looks like an older version of ROCm is installed on the system and it's choosing the incorrect version of amdclang++. I'll see if I can reproduce the issue and put in a fix.
> > 
> > 
> > I was able to reproduce and fix the previous error with hipblaslt by setting Tensile_COMPILER in the recipe:
> > ```
> > diff --git a/repos/spack_repo/builtin/packages/hipblaslt/package.py b/repos/spack_repo/builtin/packages/hipblaslt/package.py
> > index eba3fbc25b..e2337352ee 100644
> > --- a/repos/spack_repo/builtin/packages/hipblaslt/package.py
> > +++ b/repos/spack_repo/builtin/packages/hipblaslt/package.py
> > @@ -278,6 +278,11 @@ class Hipblaslt(CMakePackage):
> >                      "ROCROLLER_ASSEMBLER_PATH", f"{self.spec['llvm-amdgpu'].prefix}/bin/amdclang++"
> >                  )
> >              )
> > +            args.append(
> > +                self.define(
> > +                    "Tensile_COMPILER", f"{self.spec['llvm-amdgpu'].prefix}/bin/amdclang++"
> > +                )
> > +            )
> >          if self.spec.satisfies("@7.1:"):
> >              args.append(self.define("HIPBLASLT_ENABLE_CLIENT", self.run_tests))
> >          else:
> > ```
> 
> Good suggestion, thank you 👍 Any reason why this was not merged in `develop`?

I was trying to get my 7.1.0 PR merged for the past couple weeks, I didn't want to add any additional changes because I thought it might delay it getting merged. We'll add the change with 7.1.1: https://github.com/spack/spack-packages/pull/2782

iomaganaris committed:
Fix 7.1.0 hipblaslt

iomaganaris committed:
Fix compiler version in yml file

iomaganaris commented:
cscs-ci run alps;system=beverin;uarch=mi200;uenv=prgenv-gnu:25.12

iomaganaris committed:
New try with fixes of clang-offload-bunder

iomaganaris committed:
Disable all the rocprofiler related packages

iomaganaris commented:
cscs-ci run alps;system=beverin;uarch=mi200;uenv=prgenv-gnu:25.12

iomaganaris committed:
Add unwind dependency to rocprofiler-register and use python 3.12.12 with llvm-amdgpu to avoid rebuilding python packages

simonpintarelli committed:
cray-mpich@8.1.33

simonpintarelli commented:
cscs-ci run alps;system=beverin;uarch=mi200;uenv=prgenv-gnu:25.12

simonpintarelli committed:
cray-mpich@8.1.33 in rccl-test dependencies

simonpintarelli commented:
cscs-ci run alps;system=beverin;uarch=mi200;uenv=prgenv-gnu:25.12

simonpintarelli commented:
@msimberg I built q-e-sirius with openmpi+rocm7:
https://github.com/eth-cscs/alps-uenv/pull/288/files

 I'm not sure how to run it correctly:

```
srun --mpi=pmix --jobid=205659 --overlap -n2 osu_bw D D                                                                                                    
--------------------------------------------------------------------------                                                                                                                                    
A requested component was not found, or was unable to be opened.  This                                                                                                                                        
means that this component is either not installed or is unable to be                                                                                                                                          
used on your system (e.g., sometimes this means that shared libraries                                                                                                                                         
that the component requires are unable to be found/loaded).  Note that                                                                                                                                        
PMIx stopped checking at the first component that it did not find.                                     

Host:      nid002920                                                                                                                                                                                          
Framework: psec                                                                                                                                                                                               
Component: munge  
```
`--mpi=pmi2` worked, but still gave a warning:
```
[beverin][simonpi@nid002920 osu-micro-benchmarks]$ srun --mpi=pmi2 --jobid=205659 --overlap -n2 osu_bw D D
No PMIx server was reachable, but a PMI1/2 was detected.
If srun is being used to launch application,  2 singletons will be started.
```

msimberg commented:
@simonpintarelli have a look here: https://docs.tds.cscs.ch/301/software/communication/openmpi/#uenv (not merged yet). The munge warning should be harmless if it ran otherwise (but I'm pretty sure you need to set the other variables mentioned on that page). That said, pmix might be set up differently on beverin and I haven't tested there.

CXI should be the safe choice. LNX is faster if it works, but it may not work...

iomaganaris committed:
Trying to enable rocprofiler-compute and -systems

iomaganaris committed:
Update py-dash recipe

iomaganaris commented:
cscs-ci run alps;system=beverin;uarch=mi200;uenv=prgenv-gnu:25.12

iomaganaris committed:
Trying to install all the profilers and fix issue with hipsparse not linked to rocsparse

iomaganaris committed:
Expose hsa-rocr-dev package and add miopen-hip

iomaganaris commented:
cscs-ci run alps;system=beverin;uarch=mi200;uenv=prgenv-gnu:25.12

iomaganaris commented:
cscs-ci run alps;system=beverin;uarch=mi200;uenv=prgenv-gnu:25.12

iomaganaris committed:
Increase build CI timelimit

iomaganaris committed:
Try building hipsparse as static library to avoid issues in cupy

Issue is:
ImportError: /user-environment/env/default/lib/libhipsparse.so.4: undefined symbol: rocsparse_destroy_spmv_descr

which is strange because:
 ldd /user-environment/env/default/lib/libhipsparse.so.4
        linux-vdso.so.1 (0x00007ffddc1c1000)
        /user-environment/env/default/lib/librocsparse.so (0x00007f6af9a27000)
 nm /user-environment/linux-zen3/rocsparse-7.1.0-qfflrfxfu2vaqu52hitvwyukdwq767rf/lib/librocsparse.so | grep rocsparse_destroy_spmv_descr
0000000002f33fc0 T rocsparse_destroy_spmv_descr

simonpintarelli commented:
`hipblaslt` etc are in the cache now.

`composable-kernels`, a dependency of `miopen-hip` seems to take forever to build, it ran into walltime limit after 21 hrs..

iomaganaris committed:
Trying to fix LD_PRELOAD use in rocprofiler-compute

iomaganaris committed:
Try to fix patch paths

iomaganaris committed:
Fix patch

iomaganaris commented:
cscs-ci run alps;system=beverin;uarch=mi200;uenv=prgenv-gnu:25.12

iomaganaris commented:
cscs-ci run alps;system=beverin;uarch=mi200;uenv=prgenv-gnu:25.12

iomaganaris committed:
Add rocprof-trace-decoder library to rocprof-systems

iomaganaris committed:
Try fixing issue with rocsparse missing symbols in hipsparse

iomaganaris committed:
Enable trace decoder library

iomaganaris committed:
Fix trace decoder library path

iomaganaris committed:
Add the rocprof-trace-decoder package to the environment.yaml file to try and create a softlink to the /user-environment/env/default/lib directory

iomaganaris committed:
Merge remote-tracking branch 'origin/main' into prgenv-gnu-rocm-7

iomaganaris committed:
Revert changes regarding the timeout

iomaganaris reviewed (dismissed):
I have approved it however @msimberg please have another look before merging

msimberg reviewed (commented):
Thank you so much @iomaganaris for working on this! This looks good to me as well. If you don't mind, let's still try removing the explicit libfabric spec and revert the 25.7 changes?

Have you kept track of the custom packages here in terms of what is already upstream and what is changes that are only available here? If the latter, do you plan to upstream them? If the former, I guess updating spack-packages would get us some of those changes already, but let's not try it on this PR anymore as that might create new issues...

Finally, I guess we should deploy this as `prgenv-gnu/25.12-7.1.0:v1`? Do we want to keep the ROCm version in the version? We could also only use the ROCm version and drop the date? However, this is bikeshedding and I'm happy with `25.12-7.1.0` if noone has strong opinions on this. @iomaganaris and @simonpintarelli.

simonpintarelli commented:
> Finally, I guess we should deploy this as prgenv-gnu/25.12-7.1.0:v1? Do we want to keep the ROCm version in the version? We could also only use the ROCm version and drop the date? However, this is bikeshedding and I'm happy with 25.12-7.1.0 if noone has strong opinions on this. @iomaganaris and @simonpintarelli.

Good point. I think too the date can be dropped, the rocm version is enough.

iomaganaris committed:
Revert cray-mpich related changes

Co-authored-by: Mikael Simberg <mikael.simberg@iki.fi>

iomaganaris commented:
> CoralGemm indeed looks like a "simple" compilation failure, hopefully not too complicated.

Indeed: https://github.com/eth-cscs/cscs-reframe-tests/pull/524

iomaganaris committed:
Renamed 25.12 to 7.1.0

msimberg committed:
Update recipes/prgenv-gnu/7.1.0/amdgpu/environments.yaml

msimberg committed:
Update recipes/prgenv-gnu/7.1.0/amdgpu/environments.yaml

msimberg commented on recipes/prgenv-gnu/7.1.0/amdgpu/environments.yaml:
```diff
@@ -0,0 +1,79 @@
+gcc-env:
+  compiler: [gcc, llvm-amdgpu]
+  network:
+      mpi: cray-mpich@8.1.32+rocm
```
```suggestion
      mpi: cray-mpich@9.0.1+rocm
```

msimberg committed:
Update recipes/prgenv-gnu/7.1.0/amdgpu/environments.yaml

msimberg commented:
cscs-ci run alps;system=beverin;uarch=mi200;uenv=prgenv-gnu:7.1.0

msimberg committed:
Update recipes/prgenv-gnu/7.1.0/amdgpu/environments.yaml

msimberg committed:
Update recipes/prgenv-gnu/7.1.0/amdgpu/environments.yaml

msimberg commented:
I can't post logs now because systems are down, but yesterday I looked again a bit at the failures, and even with the patchelfed cray-mpich I get linking to libraries in `/opt/rocm/...`. However, I see the same happening on @simonpintarelli's attempt to use vanilla mpich on https://github.com/eth-cscs/alps-uenv/pull/299. E.g. `ldd $(which osu_bw)` lists completely unrelated rocm libraries in `/opt/rocm`. I noticed that if I start the modules view and just `module load osu-micro-benchmarks` ldd shows only libraries that I'd expect.

For the rocm 6.X version of the uenv I had in the past some libraries missing from the view that lead to very similar types of behaviour. As far as I can tell the missing library is there (but I don't remember 100% sure which ones it was, I think it was `comgr` or `hsa-rocr-dev` :shrug:).

msimberg committed:
Update recipes/prgenv-gnu/7.1.0/amdgpu/environments.yaml

msimberg committed:
Update recipes/prgenv-gnu/7.1.0/amdgpu/repo/packages/hipsparse/package.py

msimberg committed:
Update recipes/prgenv-gnu/7.1.0/amdgpu/environments.yaml

msimberg committed:
Update recipes/prgenv-gnu/7.1.0/amdgpu/config.yaml

msimberg committed:
Update recipes/prgenv-gnu/7.1.0/amdgpu/environments.yaml

msimberg committed:
Update recipes/prgenv-gnu/7.1.0/amdgpu/environments.yaml

msimberg committed:
Update recipes/prgenv-gnu/7.1.0/amdgpu/config.yaml

msimberg commented on recipes/prgenv-gnu/7.1.0/amdgpu/config.yaml:
```diff
@@ -0,0 +1,9 @@
+  repo: https://github.com/spack/spack.git
+  packages:
+    repo: https://github.com/msimberg/spack-packages.git
+    commit: b49741c7c536246ca95762193100907befd06214 # develop on 2025-12-19 + https://github.com/spack/spack-packages/compare/develop...msimberg:spack-packages:aws-ofi-nccl-with-rocm
```
```suggestion
    commit: aws-ofi-nccl-with-rocm # develop on 2025-12-19 + https://github.com/spack/spack-packages/compare/develop...msimberg:spack-packages:aws-ofi-nccl-with-rocm
```

msimberg committed:
Update recipes/prgenv-gnu/7.1.0/amdgpu/config.yaml

msimberg commented on recipes/prgenv-gnu/7.1.0/amdgpu/environments.yaml:
```diff
@@ -0,0 +1,78 @@
+      uenv:
+        add_compilers: true
+        prefix_paths:
+          LD_LIBRARY_PATH: [lib, lib64]
```
```suggestion
          LD_LIBRARY_PATH: [lib, lib64]
        env_vars:
          set:
          - NCCL_NET_PLUGIN: "ofi"
          - HSA_NO_SCRATCH_RECLAIM: "1"
```
see https://github.com/eth-cscs/alps-uenv/pull/299/files#r2912985599.

msimberg committed:
Update recipes/prgenv-gnu/7.1.0/amdgpu/environments.yaml

msimberg commented:
cscs-ci run alps;system=beverin;uarch=mi200;uenv=prgenv-gnu:7.1.0

msimberg commented:
This build also seems to be working a lot better (similar to the vanilla mpich build in https://github.com/eth-cscs/alps-uenv/pull/299#issuecomment-4033371242) after removing the LD_PRELOAD, adding https://github.com/eth-cscs/alps-cluster-config/pull/79, and updating aws-ofi-nccl, but not perfectly.

RCCL tests have slightly worse performance on mi200 but that's ok and we can update the references in cscs-reframe-tests.

HPL is still failing and I have no idea what's going on there.

iomaganaris committed:
Add netcdf and revert change to LD_PRELOAD to create usable uenv

iomaganaris commented:
cscs-ci run alps;system=beverin;uarch=mi200;uenv=prgenv-gnu:7.1.0

iomaganaris commented:
I reverted commenting out `LD_PRELOAD` because I need a new uenv with `netcdf` 🙈

msimberg committed:
Disable LD_PRELOAD hack again

msimberg committed:
Use cray-mpich 9.1

msimberg commented:
cscs-ci run alps;system=beverin;uarch=mi200;uenv=prgenv-gnu:7.1.0

msimberg commented:
> I reverted commenting out `LD_PRELOAD` because I need a new uenv with `netcdf` 🙈

No worries :P I commented it out again (but feel free to reenable it again if you need a new build). I updated the recipe to use the new cray-mpich 9.1 which should officially be meant for rocm 7 (thanks @simonpintarelli for updating it!).

msimberg commented on recipes/prgenv-gnu/7.1.0/amdgpu/config.yaml:
```diff
@@ -0,0 +1,9 @@
+  repo: https://github.com/spack/spack.git
+  packages:
+    repo: https://github.com/msimberg/spack-packages.git
+    commit: aws-ofi-nccl-with-rocm # develop on 2025-12-19 + https://github.com/spack/spack-packages/compare/develop...msimberg:spack-packages:aws-ofi-nccl-with-rocm
```
```suggestion
    commit: aws-ofi-nccl-with-rocm-old-develop # develop on 2025-12-19 + https://github.com/spack/spack-packages/compare/develop...msimberg:spack-packages:aws-ofi-nccl-with-rocm-old-develop
```

msimberg committed:
Apply suggestion from @msimberg

msimberg commented:
cscs-ci run alps;system=beverin;uarch=mi200;uenv=prgenv-gnu:7.1.0

msimberg commented:
cscs-ci run alps;system=beverin;uarch=mi200;uenv=prgenv-gnu:7.1.0
